WIP: K3d SRIOV

oshoval opened the pull request.

oshoval commented:
Basic sanity passed
but there is lots of work still

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2633/rehearsal-check-up-k3d-1.25-sriov/1630148954723192832

enp0s3 commented:
/uncc
@oshoval Feel free to CC me if needed

oshoval commented:
> /uncc @oshoval Feel free to CC me if needed

Thanks, note that uncc will continue to nag emails, you can unsubscribe instead that would help
sorry for the noise, even if i uncced you it wouldn't stop the email IIRC.

oshoval commented:
/cc @EdDev @ormergi @orelmisan

oshoval commented:
/uncc @awels

oshoval commented:
Drafting because we don't need the lanes atm.
But ready to review
Thanks

oshoval commented:
/cc @vladikr 
You might be interested to use it for vpgu lane
Just need to remove the part that configure the SRIOV

EdDev commented on cluster-up/cluster/k3d/common.sh:
```diff
@@ -2,312 +2,173 @@
-    cat ${KUBEVIRTCI_CONFIG_PATH}/$KUBEVIRT_PROVIDER/kind.yaml
-}
+    NETWORK=bridge
+    if [ $CRI_BIN == podman ]; then
```
I do not fully understand.
If it is podman it pulls the image and if it is not, it does not load it?

oshoval replied:
This is the temporary WA for running podman locally
when we are using docker, docker supports short-names, i.e `fedora` which stands for docker.io/fedora
podman doesnt support it
the main server image that k3d pulls is shortended, so podman cant pull it, so we help it by that, and then all works (locally)

oshoval replied:
Removing it, with podman 4 it works

For fedora this can be used
https://podman.io/blogs/2022/03/06/why_no_podman4_f35.html

EdDev commented on cluster-up/cluster/k3d/common.sh:
```diff
@@ -2,312 +2,173 @@
+
+    for agent_node in $(_get_agent_nodes); do
+        if ip netns exec $agent_node ip -d a | grep "vf 0" -B 2 > /dev/null; then
+            iface=$(ip netns exec $agent_node ip -d a | grep "vf 0" -B 2 | grep "UP" | awk -F": " '{print $2}')
```
Can you use full names for `-d`, `a` etc... It is not clear what is it or what this part does.
All the magic numbers is also unclear to me.

EdDev reviewed (commented):
Thank you for this valuable lane!

oshoval commented on cluster-up/cluster/k3d-1.25-sriov/README.md:
```diff
@@ -1,101 +1,74 @@
-  cluster, therefore, in case there is only one PF, only one cluster can be created. Locally the actual limit currently
-  supported is two clusters.
-- In order to use `make cluster-down` please make sure the right `CLUSTER_NAME` is exported.
+Note: For the initial k3d provider, used the yaml that appears on the link above, and did step [2]
+on top of that.
```
I can remove this note maybe?

I meant that for this PR, k3d initial version, what I did is to take the manifest they are created themself
https://k3d.io/v5.0.1/usage/advanced/calico
see this
"Or you can directly use this [calico.yaml](https://k3d.io/v5.0.1/usage/advanced/calico.yaml) manifest"
and then prefix the images with `quay.io/` (as step [2] in this PR readme suggests)

oshoval commented on cluster-up/cluster/k3d-1.25-sriov/TROUBLESHOOTING.md:
````diff
@@ -16,14 +16,14 @@ Just `go get` it by running `go get k8s.io/test-infra/prow/cmd/mkpj`
 
 ```bash
-mkpj --pull-number $KUBEVIRTPRNUMBER -job pull-kubevirt-e2e-kind-k8s-sriov-1.17.0 -job-config-path github/ci/prow/files/jobs/kubevirt/kubevirt-presubmits.yaml --config-path github/ci/prow/files/config.yaml > debugkind.yaml
+mkpj --pull-number $KUBEVIRT_PR_NUMBER -job pull-kubevirt-e2e-k3d-1.25-sriov -job-config-path github/ci/prow/files/jobs/kubevirt/kubevirt-presubmits.yaml --config-path github/ci/prow/files/config.yaml > debugkind.yaml
````
I don't think there is a script that uses it, it is more readable this way imo, i can revert if needed

EdDev replied:
ok

oshoval commented on cluster-up/cluster/k3d-1.25-sriov/sriov-node/configure_vfs.sh:
```diff
@@ -1,10 +1,10 @@
 function configure_vf_driver() {
-  local -r vf_sys_device=$1
-  local -r driver=$2
+  local vf_sys_device=$1
```
k3d has /bin/sh, not bash, this flag is not supported there

oshoval commented:
Removed podman hack for local running, podman v4 is required.
Fixed cluster-down for podman, -f is needed because the order the things are deleted 
```
WARN[0005] Failed to delete cluster network 'k3d-sriov': 'docker failed to remove network 'k3d-sriov': Error response from daemon: "k3d-sriov" has associated containers with it. Use -f to forcibly delete containers and pods: network is being used' 
```

oshoval commented:
Added commit to fetch the right kubectl in order to fix kubevirt warning
"WARNING: version difference between client (1.22)
and server (1.25) exceeds the supported minor version skew of +/-1"

oshoval commented:
@EdDev 
wdyt about renaming the provider to `k3d-sriov`, this way we dont need to rename each time we change k8s version
in case we will need two at the same time (stable and development, we can create a temp new one that does have the name), and once it replaces the old one, we can just replace the folders and the main jobs wont need to be touched
(we will just need to remove the development jobs that pointed to k3d-XYZ-sriov)

oshoval commented:
squashed the commits and did few cosmetics

EdDev commented:
> @EdDev wdyt about renaming the provider to `k3d-sriov`, this way we dont need to rename each time we change k8s version in case we will need two at the same time (stable and development, we can create a temp new one that does have the name), and once it replaces the old one, we can just replace the folders and the main jobs wont need to be touched (we will just need to remove the development jobs that pointed to k3d-XYZ-sriov)

I think it is used to keep the existing format of other jobs defined.
It is also transparent in the version used, i.e. there is no need to look into the scripts to understand what version of k8s it runs on.
I would keep it for the above reasons.

EdDev reviewed (approved):
This looks very nice, thank you!

I have one last request: Can you please add to the README the procedure to follow (and where to change it) to upgrade the components (e.g. when we will need to move to k8s 1.26).

Assume that the ones who can do it is a new memeber that just follows the instructions like a bot.

If you prefer, you can do this as a follow up.

oshoval commented:
> This looks very nice, thank you!
> 
> I have one last request: Can you please add to the README the procedure to follow (and where to change it) to upgrade the components (e.g. when we will need to move to k8s 1.26).
> 
> Assume that the ones who can do it is a new memeber that just follows the instructions like a bot.
> 
> If you prefer, you can do this as a follow up.

Thanks
Sure will add

oshoval commented:
Added info about bumping moving parts

EdDev reviewed (approved):
Thank you!

It will be nice as a follow up to drop the use of Calico, as we do not really need it.

oshoval commented:
> Thank you!
> 
> It will be nice as a follow up to drop the use of Calico, as we do not really need it.

It is problematic, there is internal flannel, that even if you want flannel, it need to be undeployed first and redeployed
because we are using multus and the internal one settings doesn't go well with our multus.
It is doable but will need deeper investigation on how to deploy flannel such that it will work with multus.
https://github.com/k8snetworkplumbingwg/multus-cni/issues/564
https://github.com/k8snetworkplumbingwg/multus-cni/issues/549
didnt work for me out of the box yet, so i used the suggested calico and it worked great, at least for now

oshoval commented:
> Thank you!
> 
> It will be nice as a follow up to drop the use of Calico, as we do not really need it.

With this https://github.com/kubevirt/kubevirtci/commit/5ab5de39d940c6af47f6de030091221bc01257b5
all the tests are passing
but need to mirror the images first and i want to validate few things
we can do it on a follow PR
Note that it isn't faster, it is lighter, but for example if we will want to use k3d for development of more non SR-IOV tests,
calico might be better isn't it?

oshoval commented:
> > With this [5ab5de3](https://github.com/kubevirt/kubevirtci/commit/5ab5de39d940c6af47f6de030091221bc01257b5)
> > all the tests are passing
> > but need to mirror the images first and i want to validate few things
> > we can do it on a follow PR
> > Note that it isn't faster, it is lighter, but for example if we will want to use k3d for development of more non SR-IOV tests,
> > calico might be better isn't it?
> 
> I do not know why Callico will be better. What feature of it are we using that requires it? My interest of dropping it is to depend on less components and having a simple setup.

dual stack ? not sure, maybe flannel also has it
we must use flannel instead if so and install it manually because of the issues mentioned above

qinqon commented on cluster-up/cluster/k3d-1.25-sriov/README.md:
```diff
@@ -0,0 +1,100 @@
+### Bumping CNI
+Update `CNI_VERSION` (see `cluster-up/cluster/k3d/common.sh` for more info)
+
+### Bumping Multus
+Download the newer manifest `https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/deployments/multus-daemonset-crio.yml`
+replace this file `cluster-up/cluster/$KUBEVIRT_PROVIDER/sriov-components/manifests/multus/multus.yaml`
+and update the kustomization file `cluster-up/cluster/$KUBEVIRT_PROVIDER/sriov-components/manifests/multus/kustomization.yaml`
+according needs.
```
Why do we need this ? We are already deploying multus with CNAO at kubevirtci

oshoval replied:
For SRIOV we deploy multus without CNAO
If we want to do refactor we can do it on another PR please,
especially where there is an ongoing effort to present multus V3 as internal script (CNAO deploys v4).

qinqon commented on cluster-up/cluster/k3d-1.25-sriov/README.md:
```diff
@@ -0,0 +1,100 @@
+and update the kustomization file `cluster-up/cluster/$KUBEVIRT_PROVIDER/sriov-components/manifests/multus/kustomization.yaml`
+according needs.
+
+### Bumping calico
+1. Fetch new calico yaml (https://docs.tigera.io/calico/3.25/getting-started/kubernetes/k3s/quickstart)
+   Enable `allow_ip_forwarding` (See https://k3d.io/v5.4.7/usage/advanced/calico)
+   Or use the one that is suggested here https://k3d.io/v5.4.7/usage/advanced/calico whenever it is updated.
+2. Prefix the images in the yaml with `quay.io/` unless they have it already.
+3. Update `cluster-up/cluster/k3d/manifests/calico.yaml` (see `CALICO` at `cluster-up/cluster/k3d/common.sh` for more info)
```
Do we need calico for this ?

oshoval replied:
yes, alternative is to use flannel, but it must be deployed manually, else multus doesn't work
problem with flannel is that we need to mirror the images to quay, where on calico we have them already.
See please

https://github.com/kubevirt/kubevirtci/pull/972#issuecomment-1456033948
https://github.com/kubevirt/kubevirtci/pull/972#issuecomment-1456154951

qinqon replied:
We can even check at follow PR using kindnet at the end is a CNI and is simpler

oshoval replied:
sounds good

qinqon commented on cluster-up/cluster/k3d-1.25-sriov/config_sriov_cluster.sh:
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
```
Nothing really here is specific to  `1.25` can we move it to a common `k3d` directory ? if we see is not the case with future k3d versions we can duplicate.

oshoval replied:
I  prefer to keep in this PR the file structure the same it was with kind, later on we can refactor
it answers also 
`First pass, in general can we move the sriov things to cluster-up/cluster/k3d/sriov.sh ?`

qinqon commented on cluster-up/cluster/k3d-1.25-sriov/provider.sh:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
Same here let's just move this to cluster-up/cluster/k3d/sriov.sh

oshoval replied:
same as https://github.com/kubevirt/kubevirtci/pull/972#discussion_r1129174222

qinqon reviewed (changes requested):
First pass, in general can we move the sriov things to `cluster-up/cluster/k3d/sriov.sh` ?

qinqon commented:
/lgtm
/approve
